throughput/pacer.go

```diff
@@ -9,54 +9,70 @@ import (
 
 	"github.com/ava-labs/hypersdk/api/ws"
 	"github.com/ava-labs/hypersdk/chain"
-	"github.com/ava-labs/hypersdk/utils"
 )
 
 type pacer struct {
 	ws *ws.WebSocketClient
 
 	inflight chan struct{}
-	done     chan error
+	done     chan struct{}
+
+	err error
 }
 
 func (p *pacer) Run(ctx context.Context, max int) {
 	p.inflight = make(chan struct{}, max)
-	p.done = make(chan error, 1)
+	p.done = make(chan struct{})
 
-	// Start a goroutine to process transaction results
-	for range p.inflight {
-		txID, result, err := p.ws.ListenTx(ctx)
-		if err != nil {
-			p.done <- fmt.Errorf("error listening to tx %s: %w", txID, err)
-			return
-		}
-		if result == nil {
-			p.done <- fmt.Errorf("tx %s expired", txID)
-			return
-		}
-		if !result.Success {
-			// Should never happen
-			p.done <- fmt.Errorf("tx failure %w: %s", ErrTxFailed, result.Error)
+	defer close(p.done)
+
+	for {
+		select {
+		case _, ok := <-p.inflight:
+			if !ok {
+				return
+			}
+			txID, result, err := p.ws.ListenTx(ctx)
+			if err != nil {
+				p.err = fmt.Errorf("error listening to tx %s: %w", txID, err)
+				return
+			}
+			if result == nil {
+				p.err = fmt.Errorf("tx %s expired", txID)
+				return
+			}
+			if !result.Success {
+				// Should never happen
+				p.err = fmt.Errorf("tx failure %w: %s", ErrTxFailed, result.Error)
+				return
+			}
+		case <-p.done:
 			return
 		}
 	}
-	p.done <- nil
 }
 
 func (p *pacer) Add(tx *chain.Transaction) error {
-	if err := p.ws.RegisterTx(tx); err != nil {
-		return err
+	select {
+	case <-p.done:
+		return p.err
+	default:
+		if err := p.ws.RegisterTx(tx); err != nil {
+			return err
+		}
```

Comment:
```suggestion
	// If Run failed, return the first error immediately, otherwise register the next tx
	select {
	case <-p.done:
		return p.err
	default:
		if err := p.ws.RegisterTx(tx); err != nil {
			return err
		}
```